mani_skill/envs/tasks/digital_twins/push_T.py

```diff
@@ -0,0 +1,509 @@
+                )
+            )
+
+            # here we write some randomization code that randomizes the x, y position of the tee we 
+            #goal tee is alredy at y = -0.1 relative to robot, so we allow the tee to be only -0.2 y relative to robot arm
+            target_region_xyz[..., 0] += torch.rand(b) * (self.tee_spawnbox_xlength) + self.tee_spawnbox_xoffset
+            target_region_xyz[..., 1] += torch.rand(b) * (self.tee_spawnbox_ylength) + self.tee_spawnbox_yoffset
+
+            ## this got 100% success in 40 mil steps
+            # target_region_xyz[..., 0] += torch.rand(b) * 0.2 - 0.1
+            # target_region_xyz[..., 1] += torch.rand(b) * 0.1 + 0.1
+
+            # target_region_xyz[..., 0] += 0 * 0.2 - 0.1
+            # #target_region_xyz[..., 1] += torch.rand(b) * 0.4 - 0.2
+            # target_region_xyz[..., 1] += 1 * 0.1 + 0.1
+            ##added here to shift T toward back of table
+            ##end added
+            target_region_xyz[..., 2] = 0.04/2 + 1e-3 #this is the half thickness of the tee plus a little
+            ##rotation for pose is just random rotation around z axis
+            ##z axis rotation euler to quaternion = [cos(theta/2),0,0,sin(theta/2)]
+            q_euler_angle = torch.rand(b)*(2*torch.pi)
+            q = torch.zeros((b,4))
+            q[:,0] = (q_euler_angle/2).cos()
+            q[:,-1] = (q_euler_angle/2).sin()
+            # we can then create a pose object using Pose.create_from_pq to then set the cube pose with. Note that even though our quaternion
+            # is not batched, Pose.create_from_pq will automatically batch p or q accordingly
+            # furthermore, notice how here we do not even using env_idx as a variable to say set the pose for objects in desired
+            # environments. This is because internally any calls to set data on the GPU buffer (e.g. set_pose, set_linear_velocity etc.)
+            # automatically are masked so that you can only set data on objects in environments that are meant to be initialized
+            obj_pose = Pose.create_from_pq(p=target_region_xyz, q=q)
+            self.tee.set_pose(obj_pose)
+
+            # here we set the location of that red/white target (the goal region). In particular here, we set the position to be in front of the cube
+            # and we further rotate 90 degrees on the y-axis to make the target object face up
+
+            xyz = torch.zeros((b, 3))
+            xyz[:] = self.ee_starting_pos2D
+            self.ee_goal_pos.set_pose(
+                Pose.create_from_pq(
+                    p=xyz,
+                    q=euler2quat(0, np.pi / 2, 0),
+                )
+            )
+
+    def evaluate(self):
+        ##success is where the overlap is over intersection thresh and ee dist to start pos is less than it's own thresh
+        inter_area = self.pseudo_render_intersection()
+        tee_place_success = (inter_area) >= self.intersection_thresh
```

Comment:
how come the intersection area is not used for dense reward computation? Did you find this to be less performant compared to the implementation you have now?

Reply:
Yes, it's far less performant. Every time I used the legacy reward function from the 2D environment, the agent found local max of pushing the vertical part of the T onto the horizontal part of the goal T to maximize intersection, and with a lack of exploration it just got stuck there (for cases where initial rotation of T was around 90/270 degrees compared to goal). Working on other finals, but will make the other fixes soon